### PR TITLE
FIXME enable 64bit bitmapset

### PR DESCRIPTION
**Pull request by yaowangm**
### Summary

Postgresql supports bitmapset uses 32 bit or 64 hit implementation. GP uses bitmapset as data store of AppendOnlyVisimapEntry. However, the implementation of visimap is based on 32bit only. The result is that bitmapset on GP is limited to 32bit, even on a 64bit platform. The PR is to enable 64bit bitmapset on 64bit platform.

### Solution

The trick is, visimap uses 32bit for not only in-memory data but also on-disk data. Code to encode/decode on-disk data (`Bitmap_Compress`/`BitmapDecompress_Decompress`) accept an uint32 array (`visiMapEntry->bitmap->words`) and encode/decode data in uint32 unit. However, `visiMapEntry->bitmap->words` is an uint64 array if bms is in 64bit.

Finally I adapted a kind of hybrid solution: 

1. Enable 64bit bitmapset (on 64bit platform), which will benefit all relevant operations.
2. visimap uses 64bit bitmapset in memory.
3. Keep encode/decode functions unchanged, i.e. they will encode/decode data in 32bit as before.

The most important thing is the join between 2 and 3. The interfaces to encode/decode on-disk data (`Bitmap_Compress`/`BitmapDecompress_Decompress`) accepts an uint32 array and a word count as input. So we can forcely convert a uint64 array to uint32 array, with doubled word count, as input. The approach can keep the encoding/decoding part unchanged, and take full advantage of 64bit bms otherwhere.

Another reason to keep the encoding/decoding part unchanged is, it works on both 32bit/64bit platform, and I don't want to implement a separated 64bit encoding/decoding part.

### Difference of on-disk data

The solution might result in some differences of on-disk data due to different data layout of 32/64 bit bms on big-endian machine. 

For example, we need to add 1 and 34 to bms, and the layout is:

On **small-endian** machine:

32bit:

range:
```
1-32        33-64
01 00 00 00 02 00 00 00
```
For 1, firstly locate the 1st word (in 32bit, range 1-32), and set the first bit (from left to right). For 34, firstly locate the 2st word (range 33-64), and set the second bit (from left to right).

64bit:

range:
```
1-64
01 00 00 00 02 00 00 00
```
For 1, firstly locate the 1st word (in 64bit, range 1-64), and set the first bit (from left to right). For 34, locate the same 1st word, and set the 34th bit (from left to right).

So there is no difference. However, it's different on **big-endian** machine:

32bit:

range:
```
1-32        33-64
00 00 00 01 00 00 00 02
```
For 1, firstly locate the 1st word (in 32bit, range 1-32), and set the first bit (from left to right). For 34, firstly locate the 2st word (range 33-64), and set the second bit (from left to right).

64bit:

range:
```
1-64
00 00 00 02 00 00 00 01
```
For 1, firstly locate the 1st word (in 64bit, range 0-63), and set the first bit (from left to right). For 34, locate the same 1st word, and set the 34th bit (from left to right).

The conclusion is, **32bit and 64bit ondisk data are the same on small-endian machine, but different on big-endian machine**. Due to the difference, GPDB7 will not be able to recognize the on-disk data generated by previous versions in 32bit.

A solution: we can convert the uint64 array to uint32 array by simply exchanging the two 32bit parts of every uint64 var, e.g. "00 00 00 02 00 00 00 01" -> "00 00 00 01 00 00 00 02". After that, we can get an exactly same uint32 array for interacting with on-disk data (more precisely, the interfaces to encode/decode on-disk data, i.e. `Bitmap_Compress`/`BitmapDecompress_Decompress`).

It would not impact performance because what we actually changed is just the order of reading/writing array (in 32bit unit):
32bit: 0 1 2 3 4 5 ...
64bit: 1 0 3 2 5 4 ...

I have implemented the algorithm, but as talked, **we are lacking a big-endian machine to test the code**. So it is still a problem for now.

A possible alternative is, we might simply avoid the problem by always **adapting 32bit bms on big-endian machine**. It is feasible since big-endian machine need a special build. We will lose performance of bms on big-endian env, but it seems acceptable since big-endian env is supposed to be seldom for GPDB. @soumyadeep2007 please let me know your opinion.

Theoretically a complete solution is to re-implement visimap by a bit-stream instead of using bms indirectly. By the approach, we can isolate the bit-stream from word length and small/big endian, and get more space for performance improvement by using more efficient technologies e.g. SIMD/AVX. We can even create a common bit-stream framework since multiple funcationalities need it (e.g. bitmap index). In spite of it's worth to do or not, it is beyond the PR.

### Bitmap index

Bitmap index also uses bitmap data structure, but it implemented an independent bit stream who uses 64bit words always:

```
/*
 * The size in bits of a hybrid run-length(HRL) word.
 * If you change this, you should change the type for
 * a HRL word as well.
 */
#define BM_HRL_WORD_SIZE        64

/* the type for a HRL word */
typedef uint64   
```
So it should not be impacted. 

### Others

In the FIXME comments Heikki also mentioned "using bitmapword for TID bitmaps, just like upastream, and eliminate tbm_bitmapword". It seems not very necessary because that tbm_bitmapword is 64bit currently, and we don't have much motivation to support 32bit feature.

### Test

The relevant tests includes 3 scenarios:

1. on 64bit env: The code didn't change any existing user interface, so the scenario has been covered by existing tests, including bitmap related UTs on src/backend/utils/misc/test.
2. on 32bit env: I forcedly compiled 32bit bms by hacking code, and issued all the tests.
3. 32bit env -> 64bit env: it means creating ondisk data on 32bit env and reading the data on 64bit env (e.g. upgrading GPDB6 to GPDB7). Since existing tests don't cover pgupgrade, I had to manually issue the tests.

### CODE REVIEW

File: src/include/nodes/bitmapset.h [hunk 1]
```diff
@@ -32,23 +32,7 @@ struct List;
...
- * appendoptimized tables during upgrade (not viable).  The goal of this fixme
- * is to explore if the performance benefit of larger bitmapwords can still be
- * availed by distinguishing on-disk bitmap usage from strictly in-memory
- * bitmap usage.  E.g. define a new type bitmapword32 and use it in
- * appendoptimized code.  On a related note, tbm_bitmapword is Greenplum
- * specific 64-bit wide type used for TID bitmaps.  Can we start using
- * bitmapword for TID bitmaps, just like upastream, and eliminate
- * tbm_bitmapword?
- */
-#if false && SIZEOF_VOID_P >= 8
```
Comment (soumyadeep2007) [inline src/include/nodes/bitmapset.h:51]: For context, this comment was added in merge commit by Asim: https://github.com/greenplum-db/gpdb-postgres-merge/commit/805d13de51f132d452029c1ef9b10c17c0212731

```
 The goal of this fixme
 * is to explore if the performance benefit of larger bitmapwords can still be
 * availed by distinguishing on-disk bitmap usage from strictly in-memory
 * bitmap usage.  E.g. define a new type bitmapword32 and use it in
 * appendoptimized code
```


There are 2 goals here:
1. We should use 64 bit bitmap word sizes for all of the backend code, since bitmapset.h is used everywhere! We want all backend code to be as efficient as possible. All our users are on 64 bit systems and we want to take advantage! So we absolutely want to get rid of this `#if false`.
2. AO/CO: We cannot use 64 bit bitmap words for the on-disk representation of visimap without writing specialized code in pg_upgrade to upgrade existing visimap tuples, which will be an expensive operation to say the least. So the question then becomes, can we use 64 bit bitmap words for the in-memory portion ONLY of the visimap? If we can, does it really give us a significant performance boost?
      * We need to measure this first. For eg. you can do a large scale experiment and run perf and compare the profiles with and without your changes.
      * If there is a significant performance boost, then and only then we can consider adding the conversion code necessary for serializing/deserializing between 64 bit and 32 bit words.
      * If it doesn't then what we can do is declare a new type in appendonly_visimap.h : `typedef uint32 bitmapword32` and use it explicitly in AO visimap code.

Comment (soumyadeep2007) [reply src/include/nodes/bitmapset.h:51]:
```
On a related note, tbm_bitmapword is Greenplum
 * specific 64-bit wide type used for TID bitmaps.  Can we start using
 * bitmapword for TID bitmaps, just like upastream, and eliminate
 * tbm_bitmapword?
```

We can't really do this as it might create an upgrade headache. It's fine as is.

**Review by soumyadeep2007 — COMMENTED** [review]
> However, I didn't implement the functionality. In pgupgrade, a full vacuum will be issued which will eliminate all the visimap data on pg_aovisimap_xxx before actual upgrading. So we don't need to concern about the incompatibility of on-disk data. I am asking reviewers to pay attention to the problem and confirm the conclusion in case I missed any other possible scenario.

VACUUM FULL is not performed during pg_upgrade and visimap tables will very much be part of the upgrade set. So we have to worry about on-disk compatibility.

File: src/include/nodes/bitmapset.h (hunk 1 above)
Comment (soumyadeep2007) [reply src/include/nodes/bitmapset.h:51]: > 3\. About your suggestion: "you can do a large scale experiment and run perf and compare the profiles with and without your changes." What stuff exactly are you suggesting to test? As you said, BMS is used everywhere and we definitely need to enable 64bit BMS. I don't think it's very necessary to test the improvement of 32/64bit BMS. It also doesn't matter much for 64bit visimap, because even if we can NOT get any performance improvement of visimap, we have to do this because of the benefit of 64bit BMS, right?
> 
> For now, we have some options to solve the issue of on-disk data difference due to 32/64bit BMS:
> 
> 1. add full vacuum to pg_upgrade for all AO tables. It's pretty easy to do (see the code of `prepare_new_cluster()`), but the cost is we need more time for upgrading AO tables. It might be acceptable for a major version upgrading.
> 2. add the conversion code necessary for serializing/deserializing between 64 bit and 32 bit words ONLY for visimap ondisk based on the algorithm I showed. I believe it's feasible, and the cost is some extra complexity.
> 3. Use a special version of 32bit BMS ONLY for visimap. Note that we cannot just create a 32bit `bitmapword32` because BMS implementation depends on it at compiling time; we need to duplicate all the code of BMS. It should be very ugly. An alternative is to modify BMS implementation to make it adapt 64/32bit dynamically at runtime but not compiling time. But it should be conflicting with upstream.
> 
> I personally prefer the option 2. Please provide your comments. We can talk online if necessary. Thanks!

Option 1 is completely off the table, for the reason you mentioned. And we want to touch pg_upgrade as less as possible - we have felt agonizing pain due to differences with upstream.

Very good point about option 3.

I'm leaning towards option 2 as well. I haven't reviewed the algorithm in detail yet though. From a high level, these are the requirements:
* All in-memory bitmaps will be 64 bit (if the system is 64 bit) [including visimap in-memory struct members]
* Visimap on-disk bitmap will be 32 bit
* Serialize and deserialize code needed to transform 64-bit to 32-bit and vice-versa.
      * This code should be able to work on visimap relations that are from versions 5X/6X, as well as 7X.

**Comment by soumyadeep2007** [comment]
> A possible alternative is, we might simply avoid the problem by always adapting 32bit bms on big-endian machine. It is feasible since big-endian machine need a special build. We will lose performance of bms on big-endian env, but it seems acceptable since big-endian env is supposed to be seldom for GPDB. @soumyadeep2007 please let me know your opinion.

This sounds good to me. We don't want to be writing code that we cannot test, nor do we want to be writing a lot of extra code to support such an edge case. So, if we can detect the endianness and keep the 32-bit specific code under `ifdef`s, that would be ideal.

Please make the appropriate changes and let me know and I will take a final look.

**Comment by yaowangm** [comment]
> > A possible alternative is, we might simply avoid the problem by always adapting 32bit bms on big-endian machine. It is feasible since big-endian machine need a special build. We will lose performance of bms on big-endian env, but it seems acceptable since big-endian env is supposed to be seldom for GPDB. @soumyadeep2007 please let me know your opinion.
> 
> This sounds good to me. We don't want to be writing code that we cannot test, nor do we want to be writing a lot of extra code to support such an edge case. So, if we can detect the endianness and keep the 32-bit specific code under `ifdef`s, that would be ideal.
> 
> Please make the appropriate changes and let me know and I will take a final look.

@soumyadeep2007 Thanks for your comments. I have updated the code accordingly by adopting the solution I mentioned: disabled 64bit word bms for big-endian machine, i.e. falled back to 32bit word bms for the case, and removed  all code related to big-endian logic. Please take another look. Thanks!

BTW, it's really strange for me that GPDB supports big-endian machine but we don't have env to test. I added a TODO flag and hope we can solve the issue sometime in future.

**Review by soumyadeep2007 — COMMENTED** [review]
Given how complicated it is, I reviewed only the decompression side of things. I will follow-up on the compression side of things.

File: src/backend/access/appendonly/appendonly_visimap_entry.c [hunk 2]
```diff
@@ -129,7 +129,10 @@ AppendOnlyVisimap_GetAttrNotNull(HeapTuple t, TupleDesc td, int attr)
 void
 AppendOnlyVisiMapEnty_ReadData(AppendOnlyVisimapEntry *visiMapEntry, size_t dataSize)
 {
-	int			newWordCount;
+	/* the block count of (ondisk) bitstream */
+	int			onDiskBlockCount = 0;
+	/* the word count of in-memory bitmapset */
+	int			bmsWordCount = 0;
```
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/backend/access/appendonly/appendonly_visimap_entry.c:135]: Understood!

File: src/backend/utils/misc/bitmap_compression.c [hunk 3]
```diff
@@ -105,74 +130,11 @@ BitmapDecompress_Decompress(BitmapDecompressState *state,
 
 	if (state->compressionType == BITMAP_COMPRESSION_TYPE_NO)
 	{
-		memcpy(bitmap, 
-				Bitstream_GetAlignedData(&state->bitstream, 16), 
-				state->blockCount * sizeof(uint32));
+		Bitmap_Compress_NoDecompress(state, bitmap);
```
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/backend/utils/misc/bitmap_compression.c]: `Bitmap_Compress_PlainDecompress` doesn't really make things clearer ("plain" can mean "no" compression, "default" compression, "regular" compression etc).

File: src/backend/utils/misc/bitmap_compression.c [hunk 4]
```diff
@@ -294,47 +254,72 @@ Bitmap_Compress_Write_Header(BitmapCompressionType compressionType,
...
+	}
 
 	memset(outData, 0, maxOutDataSize);
-	blockCount = bitmapDataSize;
 
 	/* Header 
 	 */
 	Bitstream_Init(&bitstream, outData, maxOutDataSize);
-	if (!Bitmap_Compress_Write_Header(compressionType,blockCount, &bitstream))
+	if (!Bitmap_Compress_Write_Header(compressionType,
```
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/backend/utils/misc/bitmap_compression.c:311]: Yes, it is unnecessary in the context of this commit/PR. It should be done as a separate commit. All of this makes review harder.

File: src/backend/access/appendonly/appendonly_visimap_entry.c [hunk 5]
```diff
@@ -264,25 +294,66 @@ AppendOnlyVisimapEntry_GetHiddenTupleCount(
...
 	compressedBitmapSize = Bitmap_Compress(BITMAP_COMPRESSION_TYPE_DEFAULT,
-										   (visiMapEntry->bitmap ? visiMapEntry->bitmap->words : NULL),
-										   (visiMapEntry->bitmap ? visiMapEntry->bitmap->nwords : 0),
+										   (visiMapEntry->bitmap ? (uint32*)visiMapEntry->bitmap->words : NULL),
+										   BITS_PER_BITMAPWORD == 64 ?
+											bmsWordCount * 2 : bmsWordCount,
 										   visiMapEntry->data->data,
-										   bitmapSize);
+										   bitmapSize,
+										   isOnly32bitOneWord);
```
Comment (soumyadeep2007) [inline src/backend/access/appendonly/appendonly_visimap_entry.c:356]: Instead of introducing a new arg `isOnly32bitOneWord` and passing that around, could we have done this:

```
if (BITS_PER_BITMAPWORD == 64)
	{
		/*
		 * On 64bit env, if there is only one word in bms, and the last half
		 * of the word is empty, it means there is only one 32bit word actually.
		 */
		if (bmsWordCount == 1
			&& (visiMapEntry->bitmap->words[0] >> 32) == 0)
			numBlocks = 1;
		else
			numBlocks = bmsWordCount * 2;
	}
	else
		numBlocks = bmsWordCount;

	compressedBitmapSize = Bitmap_Compress(BITMAP_COMPRESSION_TYPE_DEFAULT,
										   (visiMapEntry->bitmap ? visiMapEntry->bitmap->words : NULL),
										   (visiMapEntry->bitmap ? numBlocks : 0),
										   visiMapEntry->data->data,
										   bitmapSize);
```

The idea is simply to pass `bitmapDataSize = 1` to `Bitmap_Compress`  to indicate that there is only 1 valid 32-bit word in memory. This effectively tells the compression code to ignore the 32-bit portion of all 0s - its as if it never existed!

**The rest of the code shouldn't need to change: both the memcpy and Bitmap_Compress_Default already should work as-is with 1 32-bit word? Doesn't this mean absolutely no changes to compress-side code is necessary?**

Comment (soumyadeep2007) [reply src/backend/access/appendonly/appendonly_visimap_entry.c]: IIRC bit shifting does not depend on endianness. For a 64 bit integer where the lower order bits are 0s, right-shifting by 32 should be equivalent regardless of endianness.

Besides, I would **strongly** prefer to have the simplest implementation possible. I think the current implementation is hard to follow and will be hard for any developer who wants to introduce endinanness support to follow.

Comment (yaowangm) [reply src/backend/access/appendonly/appendonly_visimap_entry.c]: OK, I gave up :-) I have updated the code by removing `isOnly32bitOneWord` and all related codes. 

Please note I still kept some basic refactorings in compress-side code because the original code is pretty mess and hard to track. I also changed a parameter name of `Bitmap_Compress()` from `bitmapDataSize` to `blockCount` because `bitmapDataSize` generally means the size of bitmap (32bit or 64bit), and `blockCount` means the the size of 32bit blocks after possible conversion.

Please take another look, thanks!

**Review by soumyadeep2007 — APPROVED** [review]
I have strong reservations against refactoring and performing a logic change in the same commit.  It not only makes it difficult for the reviewer but for anyone reading the commit while doing archaeology to isolate the logic change from the refactoring.

Also, I don't agree with the refactoring in this PR as I have pointed out multiple times. I feel the new functions added don't help and don't have clear enough interfaces. Latest example is we can't put the logic for `/* Write last RLE block */` inside `Bitmap_CompressBlock`. Functions like this which has no reuse value (only 1 caller) don't really help in reading code. Now the reader has to jump back and forth making iterative logic such as in `Bitmap_Compress_Default` really complex. Now we need a controller state and what not, whereas simple local variables sufficed before.

That said, I think the logic looks good to me and I will leave the decision to merge with you, since you have strong feelings towards the refactoring.

**Comment by yaowangm** [comment]
We decided to split the PR to a couple of sub PRs:

https://github.com/greenplum-db/gpdb/pull/14784
https://github.com/greenplum-db/gpdb/pull/14799
(TODO)

The old one will be closed.
